### ISSUE TITLE
Support debugging for tests

### ISSUE THREAD

**Issue by JasonRosenberg** [go-lang-plugin-org/go-lang-idea-plugin#1855]
I have just tried using the new delve debugger integration, which appears to be a great start.  I am using IntelliJ 141.1532, and version 0.9.491 of the plugin (which claims to be an alpha version for 1.0.0).  I am using Go 1.4.2 sdk.

Anyway, it is possible to debug files and packages that have a 'main' method.  But there does not appear to be a way yet to debug tests, via the test integration within IntelliJ.

Is this something that could be added easily enough (or are there technical reasons preventing it?)

**Comment by dlsniper** [comment]
Is it possible to debug tests with delve? If so, then this is a valid use case.

**Comment by zolotov** [comment]
@dlsniper yes, it's possible

@JasonRosenberg there are no technical reasons, just matter of time. We'll implement it eventually.

**Comment by pkieltyka** [comment]
I'm excited about this one, it's holding me back from using the debugger as I primarily write in TDD.

**Comment by pylover** [comment]
+1

**Comment by ramigg** [comment]
+1

**Comment by dlsniper** [comment]
@zolotov I'd be happy to give this one a try but I've tried to do the Go Single File configuration debugable first, as I guess the test runner should be changeable following the same principle, and I've got deeply stuck. I've managed to make dlv launch but I don't have the debugging interface active for some reason. The commit is here: https://github.com/dlsniper/google-go-lang-idea-plugin/commit/98b060eb0c13b04a13eb4e93d6f5ee595e006ff8 
If I'm not mistaken, I need to rewrite the runners so that they are built like the GoBuildingRunner on top of the AsyncGenericProgramRunner, but I'd rather not do it before I get a confirmation.

Can I have any pointers please? Thank you.

**Comment by aviadl** [comment]
+1

**Comment by ZhandosKz** [comment]
+1

**Comment by itspage** [comment]
+1

**Comment by gauravvnera** [comment]
+1

**Comment by gauravvnera** [comment]
+1

**Comment by dlsniper** [comment]
To all that you have +1-ed this, please have a look on the readme: https://github.com/go-lang-plugin-org/go-lang-idea-plugin#bumping-or-1-comments

I, we, understand that you might want a feature / bug fix done, but those comments are not helpful and may even determine (some) maintainers (not speaking for others) to not react at all. Github offers a way to be subscribed to an issue for updates, so please use that. Or please write to Github about introducing a voting feature for issues.

**Comment by jimhorng** [comment]
+1

**Comment by horkhe** [comment]
I wolud say that I much more frequently run and debug tests than applications. So it is a very important feature for people that workflow is similar to mine.

**Comment by freehan** [comment]
+1

**Comment by vvakame** [comment]
+1

**Comment by antonikonovalov** [comment]
+1

**Comment by weizhu** [comment]
+1
It seems counterintuitive to not enable debugging of tests first. In a good programming practice, developers would write and run unit tests first. Not able to debug tests make life a lot of more difficult :-(.

**Comment by msaldanha** [comment]
+1

**Comment by dlsniper** [comment]
So, before anyone else +1's this anymore, I have a working version of this, however, https://github.com/derekparker/delve/issues/423 is being a major blocker for it imo. I'll submit the PR soon for review but imho the hacks I've done there should be solved before merging.

**Comment by tnine** [comment]
Any updates on this?

**Comment by dlsniper** [comment]
@tnine not really, I currently lack the time to do the refactoring and finish it out but I'm trying to make some progress. Please follow the PR for this (and feel free to help out).

**Comment by dlsniper** [comment]
I've removed the next alpha milestone as I've been slacking off my contribution to this. I'll get back to it today.

**Comment by ramigg** [comment]
any update?

**Comment by dlsniper** [comment]
Contributions are welcomed, otherwise, no update for now, sorry.

**Comment by ramigg** [comment]
any update?

**Comment by ggeorgiev** [comment]
Lacking this feature is ... very counterproductive.  Why is this low priority?

**Comment by radius314** [comment]
👍

**Comment by sbwdlihao** [comment]
+1

**Comment by antonikonovalov** [comment]
@toothrot Why are you against?

**Comment by zolotov** [comment]
@antonikonovalov I believe he against `+1` comments which pollute his mail

**Comment by ggeorgiev** [comment]
this seems as efficient and reliable as replying to spammers to remove your email from their mailing list. There is a mute this thread link at the bottom of every email you receive - use that instead

**Comment by dlsniper** [comment]
@ggeorgiev sadly that would also mean not receiving an update when this is issue gets updated.

**Comment by zolotov** [comment]
@ggeorgiev indeed, it's really efficient and reliable but it doesn't fix all problems. E.g. I might want to receive notifications on useful comments in a particular thread and do not want to read yet another +1 comment. 
Also, +1 comments pollute the thread (not my email) and make it harder to find.

All of these are claimed here: https://github.com/dear-github/dear-github/blob/master/README.md. And emoji-reactions were introduced as an answer to the letter. So please, use emoji.

**Comment by dlsniper** [comment]
Well, the other option is to lock threads but then that's very not friendly towards the community. And if I delete all the +1 comments I get equally bad vibes for that so that's how it works for now I guess. Maybe ask Github to convert +1's to votes? :)

**Comment by zolotov** [comment]
Of course not, locking thread is really rude. I don't see a reason to do this now.

> Maybe ask Github to convert +1's to votes? :)

It was in initial letter, still not working :(

**Comment by cmartin0077** [comment]
@dlsniper, what is the best way to help with this? I tried cloning your fork but could not find any existing code for debugging a test.

**Comment by slowenthal** [comment]
I'm confused as to the status.  I see #2750 is closed.  Is there a way to grab a pre-release and debug tests?

**Comment by hnorobertson** [comment]
@slowenthal It's still WIP.
> Contributions are welcomed, otherwise, no update for now, sorry.

See commits above which reference this issue.

#2750 is simply one of many duplicates of this issue and was closed as such. It's closure has no bearing on the status of this issue.

**Comment by pylover** [comment]
This is not a real plugin. it's just cheap  monkey-patch for idea for golang

**Comment by costinm** [comment]
A quick (working for me) hack:
1. Move the dlv binary to a different location
  DLV=$HOME/.IdeaIXXX/config/plugins/Go/lib/dlv/linux/dlv 
  mv $DLV ~/bin/dlv

2. Replace it with a script:
  #!/bin/bash
  export GOPATH=.... (probably not needed)
  if [ "$6" == "test" ] ; then
    go test -c -o $4 -gcflags "-N -l" $7
    $HOME/bin/dlv $1 $2 exec $4 -- --test.v
  else
    $HOME/bin/dlv $*
  fi

3. Create a normal app config ( the app must exist ), and pass as parameters:
 test FULL_PATH_TO_TEST file

The script will take the first param (test) and build the test (replacing the app), and call
dlv. Idea and the golang plugin will believe it's debugging the app.

Breakpoints in the test work. 

Variations possible - to run individual methods, etc.

**Comment by fgblomqvist** [comment]
@costinm Could you elaborate on how to create the app config? This is what gets run if I pass it as the program arguments:
`dlv --listen=localhost:42657 --headless=true exec "/tmp/test" -- test /home/fgblomqvist/IntelliJProjects/myapp/mypackage/mypackage_test.go`

But in the end it tries to run my program with that argument and fails, since I do not have a test argument.

**Comment by costinm** [comment]
For the app config: just a dummy app ( package main, nothing inside ).
Make sure you can debug it first - than modify the launcher to add the
'test ...' params.

I suppose naming the dummy app "test_hack" and changing the script to use
that instead would
be cleaner.

What happens is that golang plugin believes it's debugging an app - and the
dlv script is
swapped so it passes different params to the real dlv.

**Comment by fgblomqvist** [comment]
Got it working, thanks! Will do until proper support is implemented. For some reason it was not able to resolve dependencies within the same package, so I had to put the tests in "mypackage_test" and import "mypackage", but after that it worked.

**Comment by strootman** [comment]
This seems like a solid work around, but I'm missing something.

I swapped out the `dlv` file with the script above, but kept having the strange permission denied errors mentioned in derekparker/delve#538.
The big "aha" moment came when I set the working directory in the run config (it defaulted to `$GOPATH`).
Then, this helped me get one step further.
I was able to create a run config, and made sure the breakpoints worked. 
But when I added the extra parameters, I get the following

      ./main_test.go:20: undefined: main

Ideas?

**Comment by costinm** [comment]
The work around has 3 steps, and it helps to add 'echo' statements to the
script to see the actual
commands/workdir/env, and do the steps manually:

1. figure out the right command to compile a test to a binary, and verify
the binary runs as a test

2. have dlv run the test binary manually

3. Trick golang plugin into believing it runs a "main" program, and make
sure the binary generated in
(1) is in the expected dir, and dlv in (2) has the expected port/flags.

The last step can be customized as you want - but before tricking the
plugin you must make sure it
works, i.e. it can compile and run a normal minimal binary. I used the
'magic' first param in config, you
can also use the program name as a way to have the dlv script change
behavior.

An interesting trick is to compile a normal binary by hand, run it - and
use the hacked "dlv" script to call
'dlv attach' ( the config will need to be edited before run to add a
"attach PID", with the pid of the
running binary).

**Comment by ryancouto** [comment]
@fgblomqvist I had the same issue, but was able to get around that by changing @costinm 's step 3 instruction to:

3. Create a normal app config ( the app must exist ), and pass as parameters:
test <package name>

**Comment by dave-r12** [comment]
@zolotov any pointers or direction for adding this to the plugin? I'd like to take a pass at it. I was going to start from this commit https://github.com/dlsniper/google-go-lang-idea-plugin/commit/101fce199d83277a6295dc0f981bccaa730cffc9

**Comment by richpoirier** [comment]
@dave-r12 Try the Gogland EAP -- debugging tests works great in there.

**Comment by dave-r12** [comment]
Oh wonderful, thanks for the heads up!

**Comment by wjx0912** [comment]
@richpoirier , good job!

**Comment by gheibia** [comment]
@richpoirier Brilliant! Although after debugging a few rounds the option was disabled.

**Comment by dlsniper** [comment]
@gheibia if you have any issues with the debugger in Gogland please open an issue here: https://youtrack.jetbrains.com/issues/Go It should not disable itself at any given time but without further details I can't help out. Thank you!

**Comment by SerkanSipahi** [comment]
The problem still exists in Gogland when you setup "edit configuration" as "Go Single File". Setup your project-file as "Go Application" :)

**Comment by dlsniper** [comment]
@SerkanSipahi how you define applications run configurations doesn't have an effect how test configurations are defined.

**Comment by SerkanSipahi** [comment]
@dlsniper sorry i forgot to complete my sentense. I think it was to late. I mean: Setup your project-file as "Go Application" will work but not when it configured as "Go Single File"

**Comment by dlsniper** [comment]
@SerkanSipahi that is correct. Only Go Application type is supported for now in the debugger, both in the plugin and Gogland. Gogland has support for debugging tests. What you are asking for is not related to tests.

**Comment by kevinleecarr** [comment]
If this is already implemented in EAP, can it get ported to the golang IDEA plugin?

**Comment by dlsniper** [comment]
The EAP and https://plugins.jetbrains.com/plugin/9568-go plugin are maintained by JetBrains. This plugin is currently not actively maintained and I have no intention to port the functionality any time in the foreseeable future.

**Comment by MEGApixel23** [comment]
Any updates?

**Comment by dlsniper** [comment]
@MEGApixel23 nobody contributed anything for this functionality in this plugin. Meanwhile JetBrains released GoLand, their Go IDE and it has full support for this. You can use the IDE from here: https://www.jetbrains.com/go/ or as a plugin in IntelliJ Ultimate (2017.3+).

**Comment by antonlisovenko** [comment]
Debugging tests in go plugin is still quite strange. I cannot evaluation most of calls ("function calls are not supported"), the inner state of maps is not showing in many situations:
![image](https://user-images.githubusercontent.com/6491995/39185979-eec75598-47c0-11e8-8ccf-02402656e93d.png)
![image](https://user-images.githubusercontent.com/6491995/39186046-1eaf4f86-47c1-11e8-9d72-6457e109bbf3.png)

I'm using the latest version of Idea and plugin (`2018.1.1 Ultimate` and `181.4445.53.182`)

**Comment by dlsniper** [comment]
Hi @antonlisovenko, this is the wrong place to report issues about the Go plugin maintained by JetBrains. Please report issues in the future here: https://youtrack.jetbrains.com/issues/Go

As for the particular issues you mentioned:
- ` Cannot evaluate functions `: this is a limitation of the Go runtime, please follow https://github.com/derekparker/delve/issues/119 and https://github.com/golang/go/issues/21678 for updates
- the ` Not expandable ` part happens because the object is nil but unfortunately we do not have a good representation at the moment.

**Comment by dlsniper** [comment]
@antonlisovenko I've created https://youtrack.jetbrains.com/issue/GO-5602 to specifically track the second issue. For the first issue see: https://youtrack.jetbrains.com/issue/GO-3433. Please vote/start them go get updates for them. Thank you.
